### PR TITLE
MONGOID-5488 - Bugfix: Date#__evolve_time__ should use the configured zone instead of Time.local (i.e. respecting Mongoid.use_activesupport_time_zone)

### PR DESCRIPTION
**Pull request by johnnyshields**
Currently `Date#__evolve_time__` uses "local" (not "UTC") which is not affected by Mongoid configuration.

This PR changes it to correct respect the `Mongoid.use_activesupport_time_zone` setting.

Refer to MONGOID-5488

### CODE REVIEW

File: spec/mongoid/criteria/queryable/extensions/date_spec.rb
Comment (johnnyshields) [inline spec/mongoid/criteria/queryable/extensions/date_spec.rb:75]: These specs were copied nearly verbatim from `Date#__mongoize_time__`.

File: lib/mongoid/criteria/queryable/extensions/date.rb
Comment (johnnyshields) [inline lib/mongoid/criteria/queryable/extensions/date.rb:28]: I've added .to_time here in order to preserve the return time. However, it's not really necessary; AS::TimeWithZone works equivalently. `__mongoize_time__` will return either Time or AS::TimeWithZone based on Time.configured. Please let me know if you'd like me to remove this `to_time` to make it consistent with `__mongoize_time__`.

File: lib/mongoid/criteria/queryable/extensions/date.rb
Comment (johnnyshields) [inline lib/mongoid/criteria/queryable/extensions/date.rb:28]: This code is now consistent with the `__mongoize_time__` behavior. There is no issue making this return `ActiveSupport::TimeWithZone` as the Ruby driver has no problem with it. However, we could suffix this with `.to_time` to avoid changing return type, but I think it's better to just be consistent with Mongoize.

**Review by p-mongo — CHANGES_REQUESTED** [review]

File: lib/mongoid/criteria/queryable/extensions/date.rb
Comment (johnnyshields) [reply lib/mongoid/criteria/queryable/extensions/date.rb:28]: AFAIK this code is correct as-is. Note `Time.configured` is a Mongoid-defined method which handles the branching for `use_activesupport_time_zone`.

It is not necessary to call `use_utc` here, as it only affects how time values are demongoized. From the Mongoid docs:

> If true, times stored in MongoDB will be returned in UTC.... The use_utc setting does not affect how times are parsed - parsing is always done in local time when the input being parsed does not include time zone information.

**Comment by johnnyshields** [comment]
@p-mongo please see my response above. Your comment is marked "changes requested", please kindly either provide further specific instructions or else mark "resolved".

**Review by p-mongo — APPROVED** [review]

**Comment by comandeo** [comment]
Is it possible that this change breaks existing code? If yes, we should at least mention this in the changelog.

**Comment by johnnyshields** [comment]
@comandeo I think it's rare, but it is certainly *possible*, so I've added the changelog. Good call.

File: docs/release-notes/mongoid-9.0.txt
Comment (p-mongo) [inline docs/release-notes/mongoid-9.0.txt:45]:
```suggestion
Use configured time zone to typecast Date to Time in queries
```

File: docs/release-notes/mongoid-9.0.txt
Comment (p-mongo) [inline docs/release-notes/mongoid-9.0.txt:46]:
```suggestion
-------------------------------------------------------------
```

**Comment by johnnyshields** [comment]
@comandeo @Neilshweky please kindly approve, thanks!

File: docs/release-notes/mongoid-9.0.txt
Comment (Neilshweky) [inline docs/release-notes/mongoid-9.0.txt:49]:
```suggestion
considers the ``Mongoid.use_activesupport_time_zone`` configuration option to perform type
```
Or something like this... Make sure the line lengths add up correctly.

**Review by Neilshweky — CHANGES_REQUESTED** [review]

File: docs/release-notes/mongoid-9.0.txt
Comment (johnnyshields) [reply docs/release-notes/mongoid-9.0.txt]: Done

**Comment by johnnyshields** [comment]
Comment addressed about. This is ready for merge.

**Review by comandeo — APPROVED** [review]

**Comment by johnnyshields** [comment]
@Neilshweky thanks for merging the other PR. This one is waiting on your approval as well.

**Review by Neilshweky — APPROVED** [review]

**Comment by johnnyshields** [comment]
thank you!

**Pull request merged**
